### PR TITLE
PHOENIX-6473 Add Hadoop JMXServlet as /jmx endpoint

### PR DESCRIPTION
**Pull request by anmolnar**
Added new Avatica server customizer to create `/jmx` endpoint powered by JMXServlet. Also had to add jetty-servlet dependency and remove shading of jetty, otherwise the Servlet class cannot be instantiated.

### CODE REVIEW

**Comment by anmolnar** [comment]
@stoty @richardantal PTAL.

File: phoenix-queryserver/src/main/java/org/apache/phoenix/queryserver/server/customizers/JMXJsonEndpointServerCustomizer.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package org.apache.phoenix.queryserver.server.customizers;
```
Comment (joshelser) [inline phoenix-queryserver/src/main/java/org/apache/phoenix/queryserver/server/customizers/JMXJsonEndpointServerCustomizer.java:1]: Needs a license header

File: phoenix-queryserver/src/main/java/org/apache/phoenix/queryserver/server/customizers/JMXJsonEndpointServerCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    Servlet servlet = new JMXJsonServlet();
+    ServletHolder holder = new ServletHolder(servlet);
+    ctx.addServlet(holder, "/");
+
+    Handler[] realHandlers = list.getChildHandlers();
+    Handler[] newHandlers = new Handler[realHandlers.length + 1];
+    newHandlers[0] = ctx;
+    System.arraycopy(realHandlers, 0, newHandlers, 1, realHandlers.length);
+    server.setHandler(new HandlerList(newHandlers));
```
Comment (joshelser) [inline phoenix-queryserver/src/main/java/org/apache/phoenix/queryserver/server/customizers/JMXJsonEndpointServerCustomizer.java:44]: Does this end up requiring authentication to access it? I don't remember if we do authentication at the Server level or in the PQS Handler itself.

**Review by joshelser — CHANGES_REQUESTED** [review]
> Also had to add jetty-servlet dependency and remove shading of jetty, otherwise the Servlet class cannot be instantiated.

What kind of error did you see when jetty-servlet is shaded? More curious than concerned (given PQS is standalone).

Need to fix the license headers, otherwise, this looks like a great addition!

File: phoenix-queryserver/src/main/java/org/apache/phoenix/queryserver/server/customizers/JMXJsonEndpointServerCustomizer.java (hunk 2 above)
Comment (anmolnar) [reply phoenix-queryserver/src/main/java/org/apache/phoenix/queryserver/server/customizers/JMXJsonEndpointServerCustomizer.java:44]: No, it doesn't need authentication (currently).

**Comment by anmolnar** [comment]
@joshelser This is the error that I see if I don't touch relocations:
```
21/05/20 17:31:58 ERROR server.QueryServer: Unrecoverable service error. Shutting down.
java.lang.NoClassDefFoundError: javax/servlet/Servlet
        at org.apache.phoenix.queryserver.server.ServerCustomizersFactory$ServerCustomizersFactoryImpl.createServerCustomizers(ServerCustomizersFactory.java:75)
```

**Comment by anmolnar** [comment]
@stoty suggested to remove the entire relocations part which caused another runtime problem:
```
javax.servlet.UnavailableException: Servlet class org.apache.hadoop.jmx.JMXJsonServlet is not a javax.servlet.Servlet
	at org.eclipse.jetty.servlet.ServletHolder.checkServletType(ServletHolder.java:555)
	at org.eclipse.jetty.servlet.ServletHolder.doStart(ServletHolder.java:385)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:749)
```

I also wanted to add an integration test to validate the new endpoint, but I see the above error no matter how I tweak relocation. To be honest shading is completely jibbrish to me, so I've given up adding a test.

**Comment by stoty** [comment]
I tried to figure out if the Jmx servlet is authenticatd in Hadoop, but I got lost in the code.
Did you check if you can access Jmx without authentication in Hadoop ?
If yes, then it's fine, if not then, then we should also require authentication here.

Also, I think that a property to optionally disable the JMX server would be useful.

**Comment by joshelser** [comment]
I forgot to reply last week, but I think this is fine to leave it unauthenticated. I can't think of anything harmful exposed via the servlet. Thanks for adding the option to disable.

**Review by joshelser — APPROVED** [review]

**Pull request closed without merging**